File: javascript/packages/linter/src/rules/html-no-style-elements.ts
import { getTagLocalName } from "@herb-tools/core"
import type { ParseResult, ParserOptions, HTMLElementNode } from "@herb-tools/core"

import { BaseRuleVisitor } from "./rule-utils.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import { ParserRule } from "../types.js"

class HTMLNoStyleElementsVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    if (getTagLocalName(node) === "style") {
      this.addOffense(
        `Avoid inline \`<style>\` tags. ${this.suggestion()}`,
        node.open_tag!.location,
      )
    }

    super.visitHTMLElementNode(node)
  }

  private suggestion(): string {
    if (this.context.framework === "actionview") {
      return "Extract the CSS into a separate `.css` file and include it with `stylesheet_link_tag`."
    }

    return "Extract the CSS into a separate `.css` file and deliver it through your framework's asset pipeline."
  }
}

export class HTMLNoStyleElementsRule extends ParserRule {
  static ruleName = "html-no-style-elements"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new HTMLNoStyleElementsVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_style_elements.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::tag_utils::{get_tag_local_name, open_tag_location};

use herb::nodes::HTMLElementNode;
use herb::ParseResult;
use herb::Visitor;
use herb_config::{Framework, Severity, SeverityConfig};

pub struct HTMLNoStyleElementsRule;

impl Rule for HTMLNoStyleElementsRule {
  fn name(&self) -> &'static str {
    "html-no-style-elements"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("unreleased")
  }

  fn default_enabled(&self) -> bool {
    false
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Error)
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      action_view_helpers: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

struct NoStyleElementsVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  framework: Option<Framework>,
}

impl NoStyleElementsVisitor {
  fn suggestion(&self) -> &'static str {
    if self.framework == Some(Framework::ActionView) {
      return "Extract the CSS into a separate `.css` file and include it with `stylesheet_link_tag`.";
    }

    "Extract the CSS into a separate `.css` file and deliver it through your framework's asset pipeline."
  }
}

impl Visitor for NoStyleElementsVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if get_tag_local_name(node).as_deref() == Some("style") {
      self.offenses.push(UnboundOffense::new(
        self.rule_name,
        format!("Avoid inline `<style>` tags. {}", self.suggestion()),
        open_tag_location(node),
      ));
    }

    self.walk_html_element_node(node);
  }
}

impl ParserRule for HTMLNoStyleElementsRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = NoStyleElementsVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      framework: context.framework,
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
